src/api/weth/WethApi.ts

```diff
@@ -0,0 +1,116 @@
+  web3: Web3
+  fetchGasPrice(): Promise<string | undefined>
+}
+
+function getWethAddressByNetwork(networkId: number): string {
+  switch (networkId) {
+    case Network.Mainnet:
+      return WETH_ADDRESS_MAINNET
+    case Network.Rinkeby:
+      return WETH_ADDRESS_RINKEBY
+    default:
+      throw new Error(`WethApi was not deployed to network ${networkId}`)
+  }
+}
+
+/**
+ * Basic implementation of WETH API
+ * Note that there's already another API for all ERC20, so only the deposit/withdraw is implemented in this API
+ */
+export class WethApiImpl implements WethApi {
+  protected _contractPrototype: WethContract
+  private web3: Web3
+  private fetchGasPrice: WethApiDependencies['fetchGasPrice']
+  protected static _contractsCache: { [network: number]: { [address: string]: WethContract } } = {}
+
+  public constructor(injectedDependencies: WethApiDependencies) {
+    Object.assign(this, injectedDependencies)
+    this._contractPrototype = new this.web3.eth.Contract(wethAbi) as WethContract
+  }
+
+  public async deposit(params: WrapUnwrapParams): Promise<Receipt> {
+    const { networkId, amount, userAddress, txOptionalParams } = params
+    const contract = await this._getContract(networkId)
+
+    const tx = contract.methods
+      .deposit()
+      .send({ from: userAddress, value: amount, gasPrice: await this.fetchGasPrice() })
+
+    if (txOptionalParams && txOptionalParams.onSentTransaction) {
+      tx.once('transactionHash', txOptionalParams.onSentTransaction)
+    }
+
+    logDebug(`[WethApi] Wrapped ${amount} ETH for user ${userAddress} in network ${networkId}`)
+
+    return tx
+  }
+
+  public async withdraw(params: WrapUnwrapParams): Promise<Receipt> {
+    const { networkId, amount, userAddress, txOptionalParams } = params
+    const contract = await this._getContract(networkId)
+
+    const tx = contract.methods.withdraw(amount).send({ from: userAddress, gasPrice: await this.fetchGasPrice() })
+
+    if (txOptionalParams && txOptionalParams.onSentTransaction) {
+      tx.once('transactionHash', txOptionalParams.onSentTransaction)
+    }
+
+    logDebug(`[WethApi] Unwapping ${amount} WETH for user ${userAddress} in network ${networkId}`)
+
+    return tx
+  }
+
+  /********************************    private methods   ********************************/
+  protected async _getContract(networkId: number): Promise<WethContract> {
+    return this._getContractForNetwork(networkId)
+  }
+
+  protected _getContractForNetwork(networkId: number): WethContract {
+    const address = getWethAddressByNetwork(networkId)
+
+    return this._getContractAtAddress(networkId, address)
+  }
+
+  protected _getContractAtAddress(networkId: number, address: string): WethContract {
+    let contract: WethContract | undefined = undefined
+
+    if (WethApiImpl._contractsCache[networkId]) {
+      contract = WethApiImpl._contractsCache[networkId][address]
+    } else {
+      WethApiImpl._contractsCache[networkId] = {}
```

Comment:
I don't get this line, if the contract doesn't exist on the `_contractsCache` why do we need to reinstantiate an empty object? is this a single item cache in which case it doesn't need to be an object?

Reply:
Now it's done differently --> by the way, is like that in the deposit.

Anyways. I think it to initialize the cache object for a network. Right after that, it adds a contract for a token.